src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -396,21 +411,55 @@ bool CodegenDataModelProvider::EndpointExists(EndpointId endpoint)
     return (emberAfIndexFromEndpoint(endpoint) != kEmberInvalidEndpointIndex);
 }
 
-EndpointId CodegenDataModelProvider::FirstEndpoint()
+std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfo(EndpointId endpoint)
+{
+    std::optional<unsigned> endpoint_idx = TryFindEndpointIndex(endpoint);
+    if (endpoint_idx.has_value())
+    {
+        return GetEndpointInfoAtIndex(static_cast<uint16_t>(*endpoint_idx));
+    }
+    return std::nullopt;
+}
+
+std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfoAtIndex(uint16_t endpointIndex)
+{
+    if (emberAfEndpointIndexIsEnabled(endpointIndex))
+    {
+        DataModel::EndpointInfo info(emberAfParentEndpointFromIndex(endpointIndex));
+        if (IsFlatCompositionForEndpointIndex(endpointIndex))
+        {
+            info.compositionPattern = DataModel::EndpointCompositionPattern::kFullFamilyPattern;
+        }
+        else if (IsTreeCompositionForEndpointIndex(endpointIndex))
+        {
+            info.compositionPattern = DataModel::EndpointCompositionPattern::kTreePattern;
+        }
+        return std::make_optional(info);
+    }
+    return std::nullopt;
+}
+
+DataModel::EndpointEntry CodegenDataModelProvider::FirstEndpoint()
 {
     // find the first enabled index
     const uint16_t lastEndpointIndex = emberAfEndpointCount();
     for (uint16_t endpoint_idx = 0; endpoint_idx < lastEndpointIndex; endpoint_idx++)
     {
         if (emberAfEndpointIndexIsEnabled(endpoint_idx))
         {
-            mEndpointIterationHint = endpoint_idx;
-            return emberAfEndpointFromIndex(endpoint_idx);
+            mEndpointIterationHint                 = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
```

Comment:
This code bit seems repeated - could we pull it out in a separate method? Should be doable in an anonymous namespace function after we move `GetEndpointInfoAtIndex` to an anon namespace as well.

Reply:
Moved the code to a function in anonymous namespace.